### ISSUE TITLE
Data reconstruction for systematic data compression

### ISSUE THREAD

**Issue by sbAsma** [clessig/atmorep#41]
to be written later

**Comment by sbAsma** [comment]
I am having the following issue: `target` isn't matching ERA5 data at the following indexes:
${lat\in [702,720]}$
${lon \in [1080,1440]}$
P.S. My current code is based on `global_forecast` mode

**Comment by clessig** [comment]
There is this error thrown? What's the stack trace?

**Comment by sbAsma** [comment]
Here is a comparison in the MSE (on normalized and denormalized data) between the old AtmoRep code (the one in the branch `main`) and the new one, in branch `develop`
old code:
```
validation loss for strategy=data_compression at epoch 0 : 0.16311338543891907
denormalized MSE = 1.3110298567584582
```
new code:
```
0: validation loss for strategy=data_compression at epoch 0 : 0.16614986956119537
0: denormalized MSE = 213.0393180847168
```
I matched the data from target to ERA5 file, and found that all data entries match except for the coordinates that I am mentioning above. And this is causing the denormalized MSE to be way too high. I am suspecting a problem in the denormalization process, what do you think @clessig ?
FYI, the only error that is raising is the rmse error

**Comment by sbAsma** [comment]
Update:
The same thing persists for global forecast evaluation mode. I printed ERA5, target, and pred. Here is a sample of the outputs for :
```
====================================== vlevel = 105 =================================================
ERA5[(0, 1439)] = 255.4578094482422
target[(0, 1439)] = 255.4578094482422
pred[(0, 1439)] = 254.625
-----------------------------------------------------------
...
-----------------------------------------------------------
ERA5[(700, 1439)] = 239.8992156982422
target[(700, 1439)] = 239.8992156982422
pred[(700, 1439)] = 238.0
-----------------------------------------------------------
ERA5[(701, 1439)] = 240.0486297607422
target[(701, 1439)] = 240.0486297607422
pred[(701, 1439)] = 238.125
-----------------------------------------------------------
ERA5[(702, 1439)] = 240.2146453857422
target[(702, 1439)] = 0.0
pred[(702, 1439)] = 0.0
-----------------------------------------------------------
ERA5[(703, 1439)] = 240.3953094482422
target[(703, 1439)] = 0.0
pred[(703, 1439)] = 0.0
-----------------------------------------------------------
...
-----------------------------------------------------------
ERA5[(720, 1439)] = 242.1277313232422
target[(720, 1439)] = 0.0
pred[(720, 1439)] = 0.0
-----------------------------------------------------------
```

**Comment by clessig** [comment]
The denormalization code has been validated. Why do you believe it is not an error that was introduced when you rebased?

**Comment by sbAsma** [comment]
> The denormalization code has been validated. Why do you believe it is not an error that was introduced when you rebased?

Ilaria has also suggested that I should investigate that, which I will be doing today.

**Comment by sbAsma** [comment]
In few days from now, I would want to perform fine tuning for the data compression/data reconstruction task. I have the following questions:

1. Which months/years of the ERA5 data set should I use?
2. Since I will be using the temperature model `yxhfkh7r`, should I just do:
    ```
    if __name__ == '__main__':
      
      try : 
    
        wandb_id, epoch, epoch_continue = 'yxhfkh7r', 69, 69
        Trainer = Trainer_BERT
        train_continue( wandb_id, epoch, Trainer, epoch_continue)
    ```
4. Aside from changing `cf.BERT_strategy ` in `train.py` to the custom mode, what else should I be changing?

**Comment by iluise** [comment]
1. years_train= list(range(1979, 2021)) and years_val = [2021]
2. yes, correct
3. Beware that in train.py you should change the BERT strategy within the train_continue() function, not in train(). Just write cf.strategy = 'xxxx' below the other overwritten properties. https://github.com/clessig/atmorep/blob/main/atmorep/core/train.py#L59

**Comment by clessig** [comment]
You should start the training first in interactive mode. The models should show some apparent convergence right at the beginning.

**Comment by sbAsma** [comment]
I need advice on how to load data when doing fine tuning on data compression. In this case, data is processed in a global manner, as opposed to BERT, which means that `shuffle()` in `MultifieldDataSampler` is no longer used.
```
  def __iter__(self):

    if self.with_shuffle :
      self.shuffle()
```
Would it make sens if I implement a method that generate random dates, and load the data according to that?

**Comment by iluise** [comment]
Hi asma, 

the global data_loading is implemented in `set_global`:
https://github.com/clessig/atmorep/blob/main/atmorep/datasets/multifield_data_sampler.py#L280

which then calls `set_data`:
https://github.com/clessig/atmorep/blob/main/atmorep/datasets/multifield_data_sampler.py#L253

my suggestion would be that you add a new method like 'data_compression` to generate the random dates in `evaluator.py` and then pass these dates to `set_global`, no? Then the masking should be done in `bert.py`, to which you will pass the new method name and where you would implement the masking. 

Maybe @clessig has a better suggestion though.

**Comment by sbAsma** [comment]
I was more inclined towards adding a method that generates random dates, and calling it inside `MultifieldDataSampler.__iter__()`

Waiting for @clessig 's input

**Comment by clessig** [comment]
If you use iluise/head for fine-tuning, you also have to use it for evaluation/inference. There are some minor details in the architecture that changed, and that's the warning you are getting (we have code to go the other way, I think, but not from the new to the old).

**Comment by sbAsma** [comment]
Hi : ) 
The table below shows data reconstruction results from the checker masking pattern:
![image](https://github.com/user-attachments/assets/6feae77e-b3a5-4dcd-866d-4b5b7c63b4a4)
- NIC: Results used for the NIC symposium paper. It was done using the old implementation of Atmorep, and the old temperature model `3qou60es`
- 3qou60es new: is same temperature model, but with the new implementation of AtmoRep.
@clessig @iluise is there an explanation for why the results are different?

**Comment by sbAsma** [comment]
> What do you mean by "new implemantation of Atmorep"? Rebasing to the current develop branch?

Yes! Is there an explanation to the difference in predictions?

**Comment by iluise** [comment]
Hi, thanks for reporting this. it looks like the same bug I'm tracking down for total_precipitation. Which commit were you using for the NIC paper? Target remains unchanged?

**Comment by sbAsma** [comment]
> Hi, thanks for reporting this. it looks like the same bug I'm tracking down for total_precipitation. Which commit were you using for the NIC paper? Target remains unchanged?

```
commit c680d90ba69a816958919c375ca330ff433c066c (old-origin/main, old-origin/HEAD)
Merge: 5c03ffe 743b699
Author: iluise <72020169+iluise@users.noreply.github.com>
Date:   Thu Feb 22 15:09:51 2024 +0100

    Merge pull request #5 from clessig/iluise-fix-dims
    
    fix dimensions
```
I am assuming target remained unchanged because I remember matching it to ERA5

**Comment by iluise** [comment]
uhm, it's a commit from February and it doesn't even have the zarr file implementation :/ so hard to say if it is a bug free commit for the pre-trained runs.. 
Are you sure that the masking scheme is masking the same things in the two commits? 

I tried to run in `global forecast` mode both with tag v0.0 and tag v1.0, that are the two reliable versions before and after the zarr implementation, and the results are very very close. what do you get if you run in `global_forecast` mode with the two codes you are comparing? 

Other debugging strategy: Can you please print down, inside the code, the model predictions for the same target (you can do this in global_forecast) in different parts of the code for the two versions you have and see where they start to diverge? starting from here:
https://github.com/clessig/atmorep/blob/main/atmorep/core/trainer.py#L380

**Comment by sbAsma** [comment]
> uhm, it's a commit from February and it doesn't even have the zarr file implementation :/ so hard to say if it is a bug free commit for the pre-trained runs.. Are you sure that the masking scheme is masking the same things in the two commits?
> 
> I tried to run in `global forecast` mode both with tag v0.0 and tag v1.0, that are the two reliable versions before and after the zarr implementation, and the results are very very close. what do you get if you run in `global_forecast` mode with the two codes you are comparing?
> 
> Other debugging strategy: Can you please print down, inside the code, the model predictions for the same target (you can do this in global_forecast) in different parts of the code for the two versions you have and see where they start to diverge? starting from here: https://github.com/clessig/atmorep/blob/main/atmorep/core/trainer.py#L380

I will try these and get back to you :wink:

**Comment by sbAsma** [comment]
Update: There was a bug when data compression masking function was transferred to the develop based code. Now that I fixed that, these are the new results:

![image](https://github.com/user-attachments/assets/0645281c-af76-4730-92a9-8e5a19d48399)

I guess there is no issue now

**Comment by clessig** [comment]
Thanks for following up.
